sklearn/ensemble/stacking_model.py

```diff
@@ -0,0 +1,173 @@
+# libraries
+import numpy as np
+
+# scikit-learn base libraries
+from sklearn.base import BaseEstimator, ClassifierMixin, RegressorMixin
+
+# scikit-learn modules
+from sklearn.cross_validation import cross_val_predict
+from sklearn.ensemble import VotingClassifier
+
+
+class StackingClassifier(BaseEstimator,ClassifierMixin):
+    '''
+    stacking ensemble classifier based on scikit-learn
+    '''
+    def __init__(self,stage_one_clfs,stage_two_clfs,weights=None, n_runs=10):
+        '''
+        
+        weights: weights of the stage_two_clfs
+        n_runs: train stage_two_clfs n_runs times and average them (only for probabilistic output)
+        '''
+        self.stage_one_clfs = stage_one_clfs
+        self.stage_two_clfs = stage_two_clfs
+        if weights == None:
+            self.weights = [1] * len(stage_two_clfs)
+        else:
+            self.weights = weights
+        self.n_runs = n_runs
+        
+    def fit(self,X,y):
+        '''
+        fit the model
+        '''
+        self.__X = X
+        self.__y = y
+        
+        # fit the first stage models
+        for clf in self.stage_one_clfs:
+            y_pred = cross_val_predict(clf, X, y, cv=5, n_jobs=1)

```

Comment:
@MechCoder I'm not sure if the paper you provide is the only (and the best) way to deal with training set partitioning for stacking.

Using CV for stage one doesn't seem to create independence issues, since each output from stage one is legit (because CV), and then fed to stage two.

Using the method described in the paper you reference sounds like doing a 2-fold CV, but stopping en-route (doing just the first fold of the CV). Except if steps 1 through 4 are meant to be looped over _n_ times, in which case it becomes CV as performed by `cross_val_predict`.

In short, using CV (manually or via `cross_val_predict` looks like the way to go here.

References:

- [Ensembles: Stacking, Super Learner - h2o Tutorial](https://github.com/h2oai/h2o-tutorials/tree/master/tutorials/ensembles-stacking#super-learner-algorithm)
- [Super Learner: Statistical Applications in Genetics and Molecular Biology](https://www.degruyter.com/view/j/sagmb.2007.6.issue-1/sagmb.2007.6.1.1309/sagmb.2007.6.1.1309.xml)